### PR TITLE
APIGW NG add Integration Response selection and parameter mapping

### PR DESCRIPTION
**Pull request by bentsku**
## Motivation

This PR starts implementing the Integration Response handler, who's responsible for selecting the right `IntegrationResponse` based on the response status code (or an `errorMessage` field very specific to the `AWS` integration with Lambda, which we will tackle later). 

The handler should then do the response parameters mapping, then the VTL rendering of the body. 
We do not test the handler yet, same as the Integration Request handler, as the behavior is incomplete yet.

Instead, we can focus on the integration response selection, and the response parameters mapping. All of the unit tests have been manually tested against AWS, with the console and a SSH tunnel with localhost.run to my own small flask server to be able to return arbitrary payload from my HTTP payload. 

More informations about the different subjects of the PR:

- https://docs.aws.amazon.com/apigateway/latest/developerguide/request-response-data-mappings.html#mapping-response-parameters
- https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-integration-settings-integration-response.html
- https://docs.aws.amazon.com/apigateway/latest/developerguide/handle-errors-in-lambda-integration.html?icmpid=apigateway_console_help

## Changes
- implement the base of the Integration Response handler, still very WIP, with comments outlining what I currently think is happening
- ported the current `selectionPattern` util to select the right integration response (thanks @dfangl for implementing it!) and unit test it
- refactor the parameter mapping a bit to share code between request and response parameter mappings (code is still pretty similar, but has a few distinct difference, we could always refactor later on)
- add the response parameter mapping logic 
- add unit tests for response parameter mapping

## TODO

What's left to do:

- [x] add unit tests for Response Parameters by validating with AWS
- [x] add unit tests for selecting the integration response

### CODE REVIEW

File: localstack-core/localstack/services/apigateway/next_gen/execute_api/parameters_mapping.py [hunk 1]
```diff
@@ -54,39 +60,145 @@ def map_integration_request(
...
                 )
 
-            elif request_mapping.startswith("'") and request_mapping.endswith("'"):
-                value = request_mapping.strip("'")
+            if value:
+                response_data_mapping["header"][header_name] = value
+
+        return response_data_mapping
+
+    def _retrieve_parameter_from_variables_and_static(
```
Comment (cloutierMat) [inline localstack-core/localstack/services/apigateway/next_gen/execute_api/parameters_mapping.py:107]: Nice! That will be ready for reuse in the gateway responses as well when we get there!

File: localstack-core/localstack/services/apigateway/next_gen/execute_api/parameters_mapping.py [hunk 2]
```diff
@@ -54,39 +60,145 @@ def map_integration_request(
...
+        :param integration_response: the Response to map parameters from
+        :return: the value to map in the ResponseDataMapping
+        """
+        if expr.startswith("body"):
+            body = integration_response.get_data() or b"{}"
+            body = body.strip()
+            try:
+                decoded_body = self._json_load(body)
+            except ValueError:
+                # x-amzn-errortype: InternalFailureException
```
Comment (cloutierMat) [inline localstack-core/localstack/services/apigateway/next_gen/execute_api/parameters_mapping.py:153]: :pray:

File: tests/unit/services/apigateway/test_parameters_mapping.py [hunk 3]
```diff
@@ -397,3 +399,228 @@ def test_default_request_mapping_casing(
...
+        mapping = mapper.map_integration_response(
+            response_parameters=response_parameters,
+            integration_response=integration_response,
+            context_variables=default_context_variables,
+            stage_variables={},
+        )
+
+        # this was validated against AWS
+        assert mapping == {
+            "header": {"body_value": "{}"},
```
Comment (cloutierMat) [inline tests/unit/services/apigateway/test_parameters_mapping.py:475]: It's back :ghost:

**Review by cloutierMat — APPROVED** [review]
:rocket: Awesome PR. Nice structure to the handler.

I still have some doubt/questions around the `AWS` lambda integration and the differences in how it handles success/error, but I guess we will discover more as we complete it's handler. For now, it seems logic to proceed this way, as the error end up in the body for the mapping, so we need it here.

Really good and clear tests. Congrats!

**Comment by bentsku** [comment]
> I still have some doubt/questions around the AWS lambda integration and the differences in how it handles success/error, but I guess we will discover more as we complete it's handler. For now, it seems logic to proceed this way, as the error end up in the body for the mapping, so we need it here.

@cloutierMat yes, it's a bit specific, but it seems it's well documented in the documentation, and it seems it really happens at the Integration Response level, as they talk about the fact that the Integration will always return `200`, success or failure. It seems like a small hacky solution on AWS side too! But we can see better later, yes. It might be that the `AWS` lambda integration might have some specificity. Surprise later 😄

**Pull request merged**